packages/sources/coinmarketcap-test/src/crypto-utils.ts

```diff
@@ -0,0 +1,261 @@
+      optionContainer = symbol
+    }
+    optionContainer.push(input)
+  })
+  if (id.length) {
+    group.id = id
+  }
+  if (slug.length) {
+    group.slug = slug
+  }
+  if (symbol.length) {
+    group.symbol = symbol
+  }
+  return Object.entries(group)
+}
+
+export const buildBatchedRequestBody = (
+  params: CryptoRequestParams[],
+  config: AdapterConfig<typeof customSettings>,
+) => {
+  //Coinmarketcap supports 3 different options for sending base params - 'id', 'slug' and 'symbol'. We group here to send batch requests for each such option. Each option should not have more than 120 unique quotes.
+  const uniqueQuotes = new Set(params.map((p) => p.quote.toUpperCase()))
+  const chunkedMatrix = chunkArray([...uniqueQuotes])
+
+  const groupList = chunkedMatrix
+    .map((chunk) => {
+      const cParams = params.filter((param) => chunk.includes(param.quote))
+      return groupByBaseOptions(cParams)
+    })
+    .flat()
+
+  return groupList.map((group) => {
+    const queryName = group[0]
+    const groupedParams = group[1]
+
+    return {
+      params: groupedParams.map((inputParams) => inputParams.payload),
+      request: {
+        baseURL: config.API_ENDPOINT,
+        url: '/cryptocurrency/quotes/latest',
+        headers: {
+          'X-CMC_PRO_API_KEY': config.API_KEY,
+        },
+        params: {
+          [queryName]: [
+            ...new Set(
+              groupedParams.map((p) => p.id || p.payload.cid || p.payload.slug || p.payload.base),
+            ),
+          ].join(','),
+          convert: [...new Set(groupedParams.map((p) => p.payload.quote))].join(','),
+        },
+      },
+    }
+  })
+}
+
+// CMC API currently uses ID as key in response, when querying with "slug" param
+const _keyForSlug = (data: ProviderResponseBody, slug: string) => {
+  if (!data || !data.data) return
+  // First try to find slug key in response (in case the API starts returning keys correctly)
+  if (Object.keys(data.data).includes(slug)) return slug
+  // Fallback to ID
+  const _iEqual = (s1: string, s2: string) => s1.toUpperCase() === s2.toUpperCase()
+  const o = Object.values(data.data).find((o) => _iEqual(o.slug, slug))
+
+  return o && o.id
+}
+
+export const constructEntry = (
+  params: CryptoRequestParams[],
+  response: ProviderResponseBody,
+  resultPath: keyof PriceInfo,
+) => {
+  return params.map((param) => {
+    let key: string | number | undefined = param.cid
+    if (!key && param.slug) {
+      key = _keyForSlug(response, param.slug)
+    }
+    if (!key) {
+      key = param.base
+    }
+
+    let result
+
+    try {
+      result = response.data[key].quote[param.quote][resultPath]
+    } catch (_) {
+      return {
+        params: param,
+        response: {
+          statusCode: 400,
+          errorMessage:
+            'Could not retrieve valid data from Data Provider. This is likely an issue with the Data Provider or the input params/overrides',
+        },
+      }
+    }
```

Comment:
I would prefer if we checked for each of the keys, that way we can return more descriptive error messages about what's wrong